src/coreclr/nativeaot/Runtime/loongarch64/AllocFast.S

```diff
@@ -0,0 +1,276 @@
+
+        // Determine whether the end of the object would lie outside of the current allocation context. If so,
+        // we abandon the attempt to allocate the object directly and fall back to the slow helper.
+        add.d  $a2, $a2, $t3
+        ld.d   $t4, $a1, OFFSETOF__Thread__m_alloc_context__alloc_limit
+        bltu  $t4, $a2, RhpNewFast_RarePath
+
+        // Update the alloc pointer to account for the allocation.
+        st.d  $a2, $a1, OFFSETOF__Thread__m_alloc_context__alloc_ptr
+
+        // Set the new objects MethodTable pointer
+        st.d  $a0, $t3, OFFSETOF__Object__m_pEEType
+
+        ori  $a0, $t3, 0
+        jirl  $r0, $ra, 0
+
+RhpNewFast_RarePath:
+        ori  $a1, $zero, 0
+        b  RhpNewObject
+    LEAF_END RhpNewFast, _TEXT
+
+// Allocate non-array object with finalizer.
+//  a0 == MethodTable
+    LEAF_ENTRY RhpNewFinalizable, _TEXT
+        ori  $a1, $zero, GC_ALLOC_FINALIZE
+        b  RhpNewObject
+    LEAF_END RhpNewFinalizable, _TEXT
+
+// Allocate non-array object.
+//  a0 == MethodTable
+//  a1 == alloc flags
+    NESTED_ENTRY RhpNewObject, _TEXT, NoHandler
+
+        PUSH_COOP_PINVOKE_FRAME $a3
+
+        // a3: transition frame
+
+        // Preserve the MethodTable in s0
+        ori  $s0, $a0, 0
+
+        addi.w  $a2, $zero, 0 // numElements
+
+        // Call the rest of the allocation helper.
+        // void* RhpGcAlloc(MethodTable *pEEType, uint32_t uFlags, uintptr_t numElements, void * pTransitionFrame)
+        bl  C_FUNC(RhpGcAlloc)
+
+        // Set the new objects MethodTable pointer on success.
+        beq  $a0, $zero, NewOutOfMemory
+
+        .cfi_remember_state
+        POP_COOP_PINVOKE_FRAME
+        EPILOG_RETURN
+
+        .cfi_restore_state
+NewOutOfMemory:
+        // This is the OOM failure path. We are going to tail-call to a managed helper that will throw
+        // an out of memory exception that the caller of this allocator understands.
+
+        ori  $a0, $s0, 0                // MethodTable pointer
+        ori  $a1, $zero, 0              // Indicate that we should throw OOM.
+
+        POP_COOP_PINVOKE_FRAME
+        b C_FUNC(RhExceptionHandling_FailedAllocation)
+
+    NESTED_END RhpNewObject, _TEXT
+
+// Allocate a string.
+//  a0 == MethodTable
+//  a1 == element/character count
+    LEAF_ENTRY RhNewString, _TEXT
+        // Make sure computing the overall allocation size wont overflow
+        lu12i.w  $a2, 0x3ffff       // (MAX_STRING_LENGTH >> 12) & 0xFFFFF
+        ori  $a2, $a2, 0xfdf        // MAX_STRING_LENGTH & 0xFFF
+        bltu  $a2, $a1, StringSizeOverflow
+
+        // Compute overall allocation size (align(base size + (element size * elements), 8)).
+        addi.w  $a2, $zero, STRING_COMPONENT_SIZE
+        ori  $a3, $zero, 29         // STRING_BASE_SIZE + 7
+        mulw.d.w  $a2, $a1, $a2
+        add.d  $a2, $a2, $a3        // a2 = (a1[31:0] * a2[31:0])[64:0] + a3
+        srli.d  $a2, $a2, 3
+        slli.d  $a2, $a2, 3
```

Comment:
Note for myself: `andi` on LA zero-extends the imm12, so unlike arm, `& -8` can't be performed in a single instruction.